libraries/AC_Avoidance/AP_OADatabase.cpp

```diff
@@ -0,0 +1,255 @@
+
+#ifndef AP_OADATABASE_SIZE_DEFAULT
+    #define AP_OADATABASE_SIZE_DEFAULT          100
+#endif
+
+#ifndef AP_OADATABASE_QUEUE_SIZE_DEFAULT
+    #define AP_OADATABASE_QUEUE_SIZE_DEFAULT 80
+#endif
+
+
+const AP_Param::GroupInfo AP_OADatabase::var_info[] = {
+
+    // @Param: SIZE
+    // @DisplayName: OADatabase maximum number of points
+    // @Description: OADatabase maximum number of points. Set to 0 to disable the OA Database.
+    // @Range: 0 10000
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("SIZE", 1, AP_OADatabase, _database_size_param, AP_OADATABASE_SIZE_DEFAULT),
+
+    // @Param: EXPIRE
+    // @DisplayName: OADatabase item timeout
+    // @Description: OADatabase item timeout. The time an item will linger without any updates before it expires. Zero means never expires which is useful for a sent-once static environment.
+    // @Units: s
+    // @Range: 0 127
+    // @Increment: 1
+    // @User: Advanced
+    AP_GROUPINFO("EXPIRE", 2, AP_OADatabase, _database_expiry_seconds, AP_OADATABASE_TIMEOUT_SECONDS_DEFAULT),
+
+    // @Param: QUEUE_SIZE
+    // @DisplayName: OADatabase queue maximum number of points
+    // @Description: OADatabase queue maximum number of points.
+    // @Range: 1 200
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("QUEUE_SIZE", 3, AP_OADatabase, _queue_size_param, AP_OADATABASE_QUEUE_SIZE_DEFAULT),
+AP_GROUPEND
+};
+
+AP_OADatabase::AP_OADatabase()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_OADatabase::init()
+{
+    init_database();
+    init_queue();
+
+    if (_queue.items == nullptr || _database.items == nullptr) {
+        gcs().send_text(MAV_SEVERITY_INFO, "DB init failed . Sizes queue:%d, db:%d", _queue.size, _database.size);
+        return;
+    }
+    gcs().send_text(MAV_SEVERITY_INFO, "DB initialised. Sizes queue:%d, db:%d", _queue.size, _database.size);
+    _initialised = true;
+}
+
+void AP_OADatabase::update()
+{
+    if (!_initialised) {
+        return;
+    }
+
+    const uint32_t now_ms = AP_HAL::millis();
+    if (now_ms - _database.optimize_last_ms > 100) {
+        _database.optimize_last_ms = now_ms;
+        optimize_db_filter();
+        database_items_remove_all_expired();
+    }
+
+    process_queue();
+}
+
+// push a location into the database
+void AP_OADatabase::queue_push(const Location &loc, const uint32_t timestamp)
+{
+    if (!_initialised) {
+        return;
+    }
+
+    const OA_DbItem item = {loc, timestamp};
+    WITH_SEMAPHORE(_queue.sem);
+    _queue.items->push(item);
+}
+
+void AP_OADatabase::init_queue()
+{
+    _queue.size = _queue_size_param;
+    if (_queue.size == 0) {
+        return;
+    }
+
+    _queue.items = new ObjectBuffer<OA_DbItem>(_queue.size);
+}
+
+void AP_OADatabase::init_database()
+{
+    _database.size = _database_size_param;
+    if (_database_size_param == 0) {
+        return;
+    }
+
+    _database.items = (OA_DbItem *)calloc(_database.size, sizeof(OA_DbItem));
```

Comment:
it looks like we will use this array even if alloc fails?

Reply:
fixed